edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs

```diff
@@ -0,0 +1,191 @@
+
+        public ReportedPropertiesStore(IEntityStore<string, TwinStoreEntity> twinStore, ICloudSync cloudSync, Option<TimeSpan> syncFrequency)
+        {
+            this.twinStore = twinStore;
+            this.cloudSync = cloudSync;
+            this.syncToCloudTask = this.SyncToCloud();
+            this.syncFrequency = syncFrequency.GetOrElse(DefaultSyncFrequency);
+        }
+
+        public async Task Update(string id, TwinCollection patch)
+        {
+            using (await this.lockProvider.GetLock(id).LockAsync())
+            {
+                Events.StoringReportedPropertiesInStore(id, patch);
+                await this.twinStore.PutOrUpdate(
+                    id,
+                    new TwinStoreEntity(patch),
+                    twinInfo =>
+                    {
+                        Events.UpdatingReportedPropertiesInStore(id, patch);
+                        TwinCollection updatedReportedProperties = twinInfo.ReportedPropertiesPatch
+                            .Map(reportedProperties => new TwinCollection(JsonEx.Merge(reportedProperties, patch, /*treatNullAsDelete*/ false)))
+                            .GetOrElse(() => patch);
+                        return new TwinStoreEntity(twinInfo.Twin, Option.Maybe(updatedReportedProperties));
+                    });
+            }
+        }
+
+        public void InitSyncToCloud(string id)
+        {
+            lock (this.syncToCloudSetLock)
+            {
+                this.syncToCloudClients.Add(id);
+            }
+
+            this.syncToCloudSignal.Set();
+        }
+
+        public async Task SyncToCloud(string id)
+        {
+            Events.SyncingReportedPropertiesToCloud(id);
+            Option<TwinStoreEntity> twinWithReportedProperties = (await this.twinStore.Get(id))
+                .Filter(t => t.ReportedPropertiesPatch.HasValue);
+            if (twinWithReportedProperties.HasValue)
+            {
+                using (await this.lockProvider.GetLock(id).LockAsync())
+                {
+                    Events.StoredReportedPropertiesFound(id);
+                    Option<TwinStoreEntity> twinInfo = await this.twinStore.Get(id);
+                    await twinInfo.ForEachAsync(
+                        async ti =>
+                        {
+                            await ti.ReportedPropertiesPatch.ForEachAsync(
+                                async reportedPropertiesPatch =>
+                                {
+                                    bool result = await this.cloudSync.UpdateReportedProperties(id, reportedPropertiesPatch);
+                                    if (result)
+                                    {
+                                        Events.UpdateReportedPropertiesSucceeded(id);
+                                        await this.twinStore.Update(
+                                            id,
+                                            t => new TwinStoreEntity(t.Twin, Option.None<TwinCollection>()));
```

Comment:
Then, once synced to the cloud, it is set to `None`?

Reply:
Right - the stored entity contains the reported properties that need to be synced to the cloud. Once that is done, the entry is cleared out. 
The entity looks like this - 
{
   Twin { ReportedProperties, DesiredProperties },
   ReportedPropertiesToSync
}
So, in this operation, the reportedPropertiesToSync is set to None.